crates/hir-ty/src/infer.rs

```diff
@@ -354,7 +354,10 @@ pub struct InferenceResult {
     /// **Note**: When a pattern type is resolved it may still contain
     /// unresolved or missing subpatterns or subpatterns of mismatched types.
     pub type_of_pat: ArenaMap<PatId, Ty>,
+    pub type_of_binding: ArenaMap<BindingId, Ty>,
     pub type_of_rpit: ArenaMap<RpitId, Ty>,
+    /// Type of the result of `.into_iter()` on the for. `ExprId` is the one of the whole for loop.
+    pub type_of_for_iterator: ArenaMap<ExprId, Ty>,
```

Comment:
This feels odd to have, though I gues we don't really have a lot of options here since we don't desugar so there is nothing to refer to for the resolver into iter type?

Reply:
Though we definitely don't want this to be an ArenaMap, as that will allocate a lot due to it being backed by a Vec. That is if we were to insert the expr id 152, we allocate a vec of length 152 where the first 151 elements are unmapped. So this should be an `FxHashMap`.

Reply:
Ah yes that's wrong. Between `FxHashMap` and `NoHashHashMap` which one would be better here?

About desugaring, it would also help in simplifing MIR lowering which currently is ugly (and this field is used there) and type inference. Currently both need to find `IntoIterator` related things via lang items, but with desugaring it would be unified in one place. But it probably has it's own cons too.

Reply:
Since this is very rarely populated with keys I'd say `NoHashHashMap`.

Though I am wondering whether we should get rid of that hashmap type again. I assumed that hashbrown has no problems with sequential hashes but I believe that was wrong (forgot where I saw that), so that type most likely has no perf gain for our usual use cases with IDs. Would need to investigate.

I was wondering about HIR desugaring in r-a, there is almost (if any at all) no desugaring happening and I am unsure whether there is a reason for that aside from the usual "there was no need for it so far".

Reply:
`NoHashHashMap` didn't work for some reason, so I used `FxHashMap`.

Experimenting with HIR desugaring can be started with `? operator`, as it would solve real problems like #12247